File: examples/bun-app/tools/schema_check.ts
#!/usr/bin/env bun
/**
 * Schema consistency checker for task template YAML vs Zod schemas.
 *
 * Compares the JSON Schema declared in task template YAML files (input_schema
 * and per-step result_schema) against the JSON Schema derived from the Zod
 * schemas that handlers actually use at runtime.
 *
 * Usage:
 *   bun run tools/schema_check.ts                    # check all templates
 *   bun run tools/schema_check.ts --verbose          # show per-field details
 *   bun run tools/schema_check.ts --template payments
 */

import { readFileSync } from "node:fs";
import { resolve, dirname } from "node:path";
import { parseArgs } from "node:util";
import { parse as parseYaml } from "yaml";
import { zodToJsonSchema } from "zod-to-json-schema";
import type { ZodTypeAny } from "zod";

import {
  // Ecommerce
  EcommerceOrderProcessingInputSchema,
  EcommerceValidateCartResultSchema,
  EcommerceProcessPaymentResultSchema,
  EcommerceUpdateInventoryResultSchema,
  EcommerceCreateOrderResultSchema,
  EcommerceSendConfirmationResultSchema,
  // Data Pipeline
  AnalyticsPipelineInputSchema,
  PipelineExtractSalesResultSchema,
  PipelineExtractInventoryResultSchema,
  PipelineExtractCustomerResultSchema,
  PipelineTransformSalesResultSchema,
  PipelineTransformInventoryResultSchema,
  PipelineTransformCustomersResultSchema,
  PipelineAggregateMetricsResultSchema,
  PipelineGenerateInsightsResultSchema,
  // Microservices
  UserRegistrationInputSchema,
  MicroservicesCreateUserResultSchema,
  MicroservicesSetupBillingActiveResultSchema,
  MicroservicesInitPreferencesResultSchema,
  MicroservicesSendWelcomeResultSchema,
  MicroservicesUpdateStatusResultSchema,
  // Customer Success
  CustomerSuccessProcessRefundInputSchema,
  CustomerSuccessValidateRefundResultSchema,
  CustomerSuccessCheckRefundPolicyResultSchema,
  CustomerSuccessApproveRefundResultSchema,
  CustomerSuccessExecuteRefundResultSchema,
  CustomerSuccessUpdateTicketResultSchema,
  // Payments
  PaymentsProcessRefundInputSchema,
  PaymentsValidateEligibilityResultSchema,
  PaymentsProcessGatewayResultSchema,
  PaymentsUpdateRecordsResultSchema,
  PaymentsNotifyCustomerResultSchema,
} from "../src/services/schemas";

// ---------------------------------------------------------------------------
// Schema registry: maps template files to their Zod schemas
// ---------------------------------------------------------------------------

interface TemplateMapping {
  yamlFile: string;
  inputSchema: ZodTypeAny;
  stepSchemas: Record<string, ZodTypeAny>;
}

const SCHEMA_REGISTRY: TemplateMapping[] = [
  {
    yamlFile: "ecommerce_order_processing.yaml",
    inputSchema: EcommerceOrderProcessingInputSchema,
    stepSchemas: {
      validate_cart: EcommerceValidateCartResultSchema,
      process_payment: EcommerceProcessPaymentResultSchema,
      update_inventory: EcommerceUpdateInventoryResultSchema,
      create_order: EcommerceCreateOrderResultSchema,
      send_confirmation: EcommerceSendConfirmationResultSchema,
    },
  },
  {
    yamlFile: "data_pipeline_analytics_pipeline.yaml",
    inputSchema: AnalyticsPipelineInputSchema,
    stepSchemas: {
      extract_sales_data: PipelineExtractSalesResultSchema,
      extract_inventory_data: PipelineExtractInventoryResultSchema,
      extract_customer_data: PipelineExtractCustomerResultSchema,
      transform_sales: PipelineTransformSalesResultSchema,
      transform_inventory: PipelineTransformInventoryResultSchema,
      transform_customers: PipelineTransformCustomersResultSchema,
      aggregate_metrics: PipelineAggregateMetricsResultSchema,
      generate_insights: PipelineGenerateInsightsResultSchema,
    },
  },
  {
    yamlFile: "microservices_user_registration.yaml",
    inputSchema: UserRegistrationInputSchema,
    stepSchemas: {
      create_user_account: MicroservicesCreateUserResultSchema,
      setup_billing_profile: MicroservicesSetupBillingActiveResultSchema,
      initialize_preferences: MicroservicesInitPreferencesResultSchema,
      send_welcome_sequence: MicroservicesSendWelcomeResultSchema,
      update_user_status: MicroservicesUpdateStatusResultSchema,
    },
  },
  {
    yamlFile: "customer_success_process_refund.yaml",
    inputSchema: CustomerSuccessProcessRefundInputSchema,
    stepSchemas: {
      validate_refund_request: CustomerSuccessValidateRefundResultSchema,
      check_refund_policy: CustomerSuccessCheckRefundPolicyResultSchema,
      get_manager_approval: CustomerSuccessApproveRefundResultSchema,
      execute_refund_workflow: CustomerSuccessExecuteRefundResultSchema,
      update_ticket_status: CustomerSuccessUpdateTicketResultSchema,
    },
  },
  {
    yamlFile: "payments_process_refund.yaml",
    inputSchema: PaymentsProcessRefundInputSchema,
    stepSchemas: {
      validate_payment_eligibility: PaymentsValidateEligibilityResultSchema,
      process_gateway_refund: PaymentsProcessGatewayResultSchema,
      update_payment_records: PaymentsUpdateRecordsResultSchema,
      notify_customer: PaymentsNotifyCustomerResultSchema,
    },
  },
];

// ---------------------------------------------------------------------------
// Types for comparison results
// ---------------------------------------------------------------------------

enum MismatchKind {
  FieldMissingInCode = "field_missing_in_code",
  FieldMissingInYaml = "field_missing_in_yaml",
  TypeMismatch = "type_mismatch",
  RequiredMismatch = "required_mismatch",
}

interface FieldMismatch {
  fieldName: string;
  kind: MismatchKind;
  detail: string;
}

interface SchemaComparisonResult {
  templateFile: string;
  schemaLocation: string;
  schemaName: string;
  mismatches: FieldMismatch[];
}

function isOk(result: SchemaComparisonResult): boolean {
  return result.mismatches.length === 0;
}

// ---------------------------------------------------------------------------
// JSON Schema type normalization
// ---------------------------------------------------------------------------

const JSON_SCHEMA_TO_CANONICAL: Record<string, string> = {
  string: "string",
  number: "number",
  integer: "number",
  boolean: "boolean",
  array: "array",
  object: "object",
  null: "null",
};

function canonicalType(jsonSchemaType: string | string[] | undefined): string {
  if (jsonSchemaType === undefined) {
    return "any";
  }
  if (Array.isArray(jsonSchemaType)) {
    const types = jsonSchemaType
      .filter((t) => t !== "null")
      .map((t) => JSON_SCHEMA_TO_CANONICAL[t] ?? t);
    return types.length === 1 ? types[0] : "any";
  }
  return JSON_SCHEMA_TO_CANONICAL[jsonSchemaType] ?? jsonSchemaType;
}

function resolveZodFieldType(
  fieldSchema: Record<string, unknown>,
  defs: Record<string, Record<string, unknown>>,
): string {
  // Direct type
  if ("type" in fieldSchema) {
    return canonicalType(fieldSchema.type as string | string[]);
  }

  // $ref to a $defs entry
  if ("$ref" in fieldSchema) {
    const refPath = fieldSchema.$ref as string;
    const refName = refPath.split("/").pop()!;
    const refSchema = defs[refName] ?? {};
    return canonicalType(refSchema.type as string | string[] | undefined);
  }

  // anyOf — used by Zod for Optional types → anyOf: [{type: T}, {type: null}]
  if ("anyOf" in fieldSchema) {
    const variants = fieldSchema.anyOf as Record<string, unknown>[];
    const nonNull = variants.filter(
      (s) => s.type !== "null" && !(Array.isArray(s.type) && s.type.length === 1 && s.type[0] === "null"),
    );
    if (nonNull.length === 1) {
      return resolveZodFieldType(nonNull[0], defs);
    }
    return "any";
  }

  return "any";
}

// ---------------------------------------------------------------------------
// Schema extraction
// ---------------------------------------------------------------------------

interface NormalizedSchema {
  fields: Record<string, string>;
  required: Set<string>;
}

function normalizeYamlSchema(schema: Record<string, unknown>): NormalizedSchema {
  const properties = (schema.properties ?? {}) as Record<string, Record<string, unknown>>;
  const required = new Set<string>((schema.required as string[]) ?? []);
  const fields: Record<string, string> = {};

  for (const [name, prop] of Object.entries(properties)) {
    fields[name] = canonicalType(prop.type as string | string[] | undefined);
  }

  return { fields, required };
}

function normalizeZodSchema(zodSchema: ZodTypeAny): NormalizedSchema {
  const jsonSchema = zodToJsonSchema(zodSchema) as Record<string, unknown>;
  const defs = (jsonSchema.$defs ?? jsonSchema.definitions ?? {}) as Record<
    string,
    Record<string, unknown>
  >;
  const properties = (jsonSchema.properties ?? {}) as Record<string, Record<string, unknown>>;
  const required = new Set<string>((jsonSchema.required as string[]) ?? []);
  const fields: Record<string, string> = {};

  for (const [name, prop] of Object.entries(properties)) {
    fields[name] = resolveZodFieldType(prop, defs);
  }

  return { fields, required };
}

// ---------------------------------------------------------------------------
// Comparison logic
// ---------------------------------------------------------------------------

function compareSchemas(
  yamlSchema: NormalizedSchema,
  zodSchema: NormalizedSchema,
): FieldMismatch[] {
  const mismatches: FieldMismatch[] = [];
  const allFields = new Set([
    ...Object.keys(yamlSchema.fields),
    ...Object.keys(zodSchema.fields),
  ]);

  for (const fieldName of [...allFields].sort()) {
    const inYaml = fieldName in yamlSchema.fields;
    const inCode = fieldName in zodSchema.fields;

    if (!inCode) {
      mismatches.push({
        fieldName,
        kind: MismatchKind.FieldMissingInCode,
        detail: `declared in YAML (${yamlSchema.fields[fieldName]}) but not in Zod schema`,
      });
      continue;
    }

    if (!inYaml) {
      mismatches.push({
        fieldName,
        kind: MismatchKind.FieldMissingInYaml,
        detail: `in Zod schema (${zodSchema.fields[fieldName]}) but not declared in YAML`,
      });
      continue;
    }

    // Both present — compare types
    const yamlType = yamlSchema.fields[fieldName];
    const codeType = zodSchema.fields[fieldName];
    if (yamlType !== "any" && codeType !== "any" && yamlType !== codeType) {
      mismatches.push({
        fieldName,
        kind: MismatchKind.TypeMismatch,
        detail: `YAML says '${yamlType}', code says '${codeType}'`,
      });
    }

    // Compare required status
    const yamlReq = yamlSchema.required.has(fieldName);
    const codeReq = zodSchema.required.has(fieldName);
    if (yamlReq && !codeReq) {
      mismatches.push({
        fieldName,
        kind: MismatchKind.RequiredMismatch,
        detail: "required in YAML but optional in code",
      });
    }
  }

  return mismatches;
}

// ---------------------------------------------------------------------------
// Main check logic
// ---------------------------------------------------------------------------

const TEMPLATES_DIR = resolve(
  dirname(new URL(import.meta.url).pathname),
  "..",
  "src",
  "config",
  "templates",
);

function getSchemaName(zodSchema: ZodTypeAny): string {
  // Attempt to derive a readable name; fall back to "ZodSchema"
  const desc = zodSchema.description;
  if (desc) return desc;
  return "ZodSchema";
}

function checkTemplate(mapping: TemplateMapping): SchemaComparisonResult[] {
  const yamlPath = resolve(TEMPLATES_DIR, mapping.yamlFile);
  const yamlContent = readFileSync(yamlPath, "utf-8");
  const template = parseYaml(yamlContent) as Record<string, unknown>;

  const results: SchemaComparisonResult[] = [];

  // Check input_schema
  if (mapping.inputSchema && template.input_schema) {
    const yamlNorm = normalizeYamlSchema(template.input_schema as Record<string, unknown>);
    const codeNorm = normalizeZodSchema(mapping.inputSchema);
    const mismatches = compareSchemas(yamlNorm, codeNorm);
    results.push({
      templateFile: mapping.yamlFile,
      schemaLocation: "input_schema",
      schemaName: mapping.inputSchema.description ?? "InputSchema",
      mismatches,
    });
  }

  // Check each step's result_schema
  const steps = (template.steps ?? []) as Record<string, unknown>[];
  const stepsByName = new Map<string, Record<string, unknown>>();
  for (const step of steps) {
    stepsByName.set(step.name as string, step);
  }

  for (const [stepName, zodSchema] of Object.entries(mapping.stepSchemas)) {
    const schemaName = zodSchema.description ?? stepName;
    const step = stepsByName.get(stepName);

    if (!step) {
      results.push({
        templateFile: mapping.yamlFile,
        schemaLocation: `step:${stepName}`,
        schemaName,
        mismatches: [
          {
            fieldName: "<step>",
            kind: MismatchKind.FieldMissingInYaml,
            detail: `step '${stepName}' not found in template YAML`,
          },
        ],
      });
      continue;
    }

    const resultSchema = step.result_schema as Record<string, unknown> | undefined;
    if (!resultSchema) {
      results.push({
        templateFile: mapping.yamlFile,
        schemaLocation: `step:${stepName}.result_schema`,
        schemaName,
        mismatches: [
          {
            fieldName: "<schema>",
            kind: MismatchKind.FieldMissingInYaml,
            detail: "no result_schema declared in template YAML",
          },
        ],
      });
      continue;
    }

    const yamlNorm = normalizeYamlSchema(resultSchema);
    const codeNorm = normalizeZodSchema(zodSchema);
    const mismatches = compareSchemas(yamlNorm, codeNorm);
    results.push({
      templateFile: mapping.yamlFile,
      schemaLocation: `step:${stepName}.result_schema`,
      schemaName,
      mismatches,
    });
  }

  return results;
}

function checkAll(templateFilter?: string): SchemaComparisonResult[] {
  const allResults: SchemaComparisonResult[] = [];
  for (const mapping of SCHEMA_REGISTRY) {
    if (templateFilter && !mapping.yamlFile.startsWith(templateFilter)) {
      continue;
    }
    allResults.push(...checkTemplate(mapping));
  }
  return allResults;
}

// ---------------------------------------------------------------------------
// Output formatting
// ---------------------------------------------------------------------------

const KIND_SYMBOLS: Record<MismatchKind, string> = {
  [MismatchKind.FieldMissingInCode]: "- CODE",
  [MismatchKind.FieldMissingInYaml]: "- YAML",
  [MismatchKind.TypeMismatch]: "~ TYPE",
  [MismatchKind.RequiredMismatch]: "~ REQ ",
};

function printReport(results: SchemaComparisonResult[], verbose: boolean): number {
  const totalSchemas = results.length;
  const failedSchemas = results.filter((r) => !isOk(r));

  if (failedSchemas.length === 0) {
    console.log(`All ${totalSchemas} schemas consistent.`);
    return 0;
  }

  // Group by template file
  const byFile = new Map<string, SchemaComparisonResult[]>();
  for (const r of results) {
    const existing = byFile.get(r.templateFile) ?? [];
    existing.push(r);
    byFile.set(r.templateFile, existing);
  }

  for (const [templateFile, fileResults] of byFile) {
    const fileFails = fileResults.filter((r) => !isOk(r));
    if (fileFails.length === 0 && !verbose) {
      continue;
    }

    console.log(`\n${templateFile}`);
    for (const r of fileResults) {
      const status = isOk(r) ? "OK" : "MISMATCH";
      const count = !isOk(r) ? ` (${r.mismatches.length} issues)` : "";
      console.log(`  ${status} ${r.schemaLocation} <-> ${r.schemaName}${count}`);

      if (!isOk(r) || verbose) {
        for (const m of r.mismatches) {
          const symbol = KIND_SYMBOLS[m.kind] ?? "?";
          console.log(`    ${symbol}  ${m.fieldName}: ${m.detail}`);
        }
      }
    }
  }

  const okCount = totalSchemas - failedSchemas.length;
  console.log(
    `\n${okCount}/${totalSchemas} schemas consistent, ${failedSchemas.length} with mismatches.`,
  );
  return failedSchemas.length;
}

// ---------------------------------------------------------------------------
// CLI entry point
// ---------------------------------------------------------------------------

function main(): void {
  const { values } = parseArgs({
    args: Bun.argv.slice(2),
    options: {
      verbose: { type: "boolean", short: "v", default: false },
      template: { type: "string", short: "t" },
    },
    strict: true,
  });

  const results = checkAll(values.template);
  const failCount = printReport(results, values.verbose ?? false);
  process.exit(failCount > 0 ? 1 : 0);
}

main();

File: examples/fastapi-app/tools/schema_check.py
"""Schema consistency checker for task template YAML vs working Pydantic types.

Compares the JSON Schema declared in task template YAML files (input_schema and
per-step result_schema) against the JSON Schema derived from the Pydantic models
that handlers actually use at runtime.

Usage:
    uv run python -m tools.schema_check                # check all templates
    uv run python -m tools.schema_check --verbose      # show per-field details
    uv run python -m tools.schema_check --template payments_process_refund

Can also be imported and called from pytest (see tests/test_schema_consistency.py).
"""

from __future__ import annotations

import argparse
import sys
from dataclasses import dataclass, field
from enum import Enum
from pathlib import Path
from typing import Any

import yaml
from pydantic import BaseModel

from app.services.types import (
    # Ecommerce
    EcommerceOrderProcessingInput,
    EcommerceValidateCartResult,
    EcommerceProcessPaymentResult,
    EcommerceUpdateInventoryResult,
    EcommerceCreateOrderResult,
    EcommerceSendConfirmationResult,
    # Data Pipeline
    DataPipelineAnalyticsPipelineInput,
    PipelineExtractSalesResult,
    PipelineExtractInventoryResult,
    PipelineExtractCustomerResult,
    PipelineTransformSalesResult,
    PipelineTransformInventoryResult,
    PipelineTransformCustomersResult,
    PipelineAggregateMetricsResult,
    PipelineGenerateInsightsResult,
    # Microservices
    MicroservicesUserRegistrationInput,
    MicroservicesCreateUserResult,
    MicroservicesSetupBillingResult,
    MicroservicesInitPreferencesResult,
    MicroservicesSendWelcomeResult,
    MicroservicesUpdateStatusResult,
    # Customer Success
    CustomerSuccessProcessRefundInput,
    CustomerSuccessValidateRefundResult,
    CustomerSuccessCheckPolicyResult,
    CustomerSuccessApproveRefundResult,
    CustomerSuccessExecuteRefundResult,
    CustomerSuccessUpdateTicketResult,
    # Payments
    PaymentsProcessRefundInput,
    PaymentsValidateEligibilityResult,
    PaymentsProcessGatewayResult,
    PaymentsUpdateRecordsResult,
    PaymentsNotifyCustomerResult,
)


# ---------------------------------------------------------------------------
# Schema registry: maps template files to their Pydantic models
# ---------------------------------------------------------------------------


@dataclass
class TemplateMapping:
    """Maps a template YAML to its input model and per-step result models."""

    yaml_file: str
    input_model: type[BaseModel] | None = None
    step_models: dict[str, type[BaseModel]] = field(default_factory=dict)


SCHEMA_REGISTRY: list[TemplateMapping] = [
    TemplateMapping(
        yaml_file="ecommerce_order_processing.yaml",
        input_model=EcommerceOrderProcessingInput,
        step_models={
            "validate_cart": EcommerceValidateCartResult,
            "process_payment": EcommerceProcessPaymentResult,
            "update_inventory": EcommerceUpdateInventoryResult,
            "create_order": EcommerceCreateOrderResult,
            "send_confirmation": EcommerceSendConfirmationResult,
        },
    ),
    TemplateMapping(
        yaml_file="data_pipeline_analytics_pipeline.yaml",
        input_model=DataPipelineAnalyticsPipelineInput,
        step_models={
            "extract_sales_data": PipelineExtractSalesResult,
            "extract_inventory_data": PipelineExtractInventoryResult,
            "extract_customer_data": PipelineExtractCustomerResult,
            "transform_sales": PipelineTransformSalesResult,
            "transform_inventory": PipelineTransformInventoryResult,
            "transform_customers": PipelineTransformCustomersResult,
            "aggregate_metrics": PipelineAggregateMetricsResult,
            "generate_insights": PipelineGenerateInsightsResult,
        },
    ),
    TemplateMapping(
        yaml_file="microservices_user_registration.yaml",
        input_model=MicroservicesUserRegistrationInput,
        step_models={
            "create_user_account": MicroservicesCreateUserResult,
            "setup_billing_profile": MicroservicesSetupBillingResult,
            "initialize_preferences": MicroservicesInitPreferencesResult,
            "send_welcome_sequence": MicroservicesSendWelcomeResult,
            "update_user_status": MicroservicesUpdateStatusResult,
        },
    ),
    TemplateMapping(
        yaml_file="customer_success_process_refund.yaml",
        input_model=CustomerSuccessProcessRefundInput,
        step_models={
            "validate_refund_request": CustomerSuccessValidateRefundResult,
            "check_refund_policy": CustomerSuccessCheckPolicyResult,
            "get_manager_approval": CustomerSuccessApproveRefundResult,
            "execute_refund_workflow": CustomerSuccessExecuteRefundResult,
            "update_ticket_status": CustomerSuccessUpdateTicketResult,
        },
    ),
    TemplateMapping(
        yaml_file="payments_process_refund.yaml",
        input_model=PaymentsProcessRefundInput,
        step_models={
            "validate_payment_eligibility": PaymentsValidateEligibilityResult,
            "process_gateway_refund": PaymentsProcessGatewayResult,
            "update_payment_records": PaymentsUpdateRecordsResult,
            "notify_customer": PaymentsNotifyCustomerResult,
        },
    ),
]


# ---------------------------------------------------------------------------
# Types for comparison results
# ---------------------------------------------------------------------------


class MismatchKind(Enum):
    FIELD_MISSING_IN_CODE = "field_missing_in_code"
    FIELD_MISSING_IN_YAML = "field_missing_in_yaml"
    TYPE_MISMATCH = "type_mismatch"
    REQUIRED_MISMATCH = "required_mismatch"


@dataclass
class FieldMismatch:
    field_name: str
    kind: MismatchKind
    detail: str


@dataclass
class SchemaComparisonResult:
    template_file: str
    schema_location: str  # e.g. "input_schema" or "step:validate_cart.result_schema"
    model_name: str
    mismatches: list[FieldMismatch] = field(default_factory=list)

    @property
    def ok(self) -> bool:
        return len(self.mismatches) == 0


# ---------------------------------------------------------------------------
# JSON Schema type normalization
# ---------------------------------------------------------------------------

# Map from JSON Schema type strings to a canonical set for comparison
_JSON_SCHEMA_TO_CANONICAL = {
    "string": "string",
    "number": "number",
    "integer": "number",  # treat integer as compatible with number
    "boolean": "boolean",
    "array": "array",
    "object": "object",
    "null": "null",
}


def _canonical_type(json_schema_type: str | list | None) -> str:
    """Normalize a JSON Schema type to a canonical string for comparison."""
    if json_schema_type is None:
        return "any"
    if isinstance(json_schema_type, list):
        types = [_JSON_SCHEMA_TO_CANONICAL.get(t, t) for t in json_schema_type if t != "null"]
        return types[0] if len(types) == 1 else "any"
    return _JSON_SCHEMA_TO_CANONICAL.get(json_schema_type, json_schema_type)


def _resolve_pydantic_field_type(
    field_schema: dict[str, Any],
    defs: dict[str, Any],
) -> str:
    """Resolve a Pydantic JSON Schema field to a canonical type string.

    Handles $ref, anyOf (for Optional types), and direct type declarations.
    """
    # Direct type
    if "type" in field_schema:
        return _canonical_type(field_schema["type"])

    # $ref to a $defs entry
    if "$ref" in field_schema:
        ref_name = field_schema["$ref"].rsplit("/", 1)[-1]
        ref_schema = defs.get(ref_name, {})
        return _canonical_type(ref_schema.get("type"))

    # anyOf — used by Pydantic for Optional[T] → anyOf: [{type: T}, {type: null}]
    if "anyOf" in field_schema:
        non_null = [s for s in field_schema["anyOf"] if s.get("type") != "null"]
        if len(non_null) == 1:
            return _resolve_pydantic_field_type(non_null[0], defs)
        # Multiple non-null options — too complex, treat as any
        return "any"

    return "any"


# ---------------------------------------------------------------------------
# Schema extraction
# ---------------------------------------------------------------------------


@dataclass
class NormalizedSchema:
    """Flattened representation of a JSON Schema's top-level fields."""

    fields: dict[str, str]  # field_name -> canonical type
    required: set[str]


def normalize_yaml_schema(schema: dict[str, Any]) -> NormalizedSchema:
    """Extract top-level fields and types from a YAML-declared JSON Schema."""
    properties = schema.get("properties", {})
    required = set(schema.get("required", []))
    fields = {}
    for name, prop in properties.items():
        fields[name] = _canonical_type(prop.get("type"))
    return NormalizedSchema(fields=fields, required=required)


def normalize_pydantic_schema(model: type[BaseModel]) -> NormalizedSchema:
    """Extract top-level fields and types from a Pydantic model's JSON Schema."""
    schema = model.model_json_schema()
    defs = schema.get("$defs", {})
    properties = schema.get("properties", {})
    required = set(schema.get("required", []))
    fields = {}
    for name, prop in properties.items():
        fields[name] = _resolve_pydantic_field_type(prop, defs)
    return NormalizedSchema(fields=fields, required=required)


# ---------------------------------------------------------------------------
# Comparison logic
# ---------------------------------------------------------------------------


def compare_schemas(
    yaml_schema: NormalizedSchema,
    pydantic_schema: NormalizedSchema,
) -> list[FieldMismatch]:
    """Compare a YAML-declared schema against a Pydantic-derived schema."""
    mismatches: list[FieldMismatch] = []

    all_fields = set(yaml_schema.fields.keys()) | set(pydantic_schema.fields.keys())

    for field_name in sorted(all_fields):
        in_yaml = field_name in yaml_schema.fields
        in_code = field_name in pydantic_schema.fields

        if not in_code:
            mismatches.append(FieldMismatch(
                field_name=field_name,
                kind=MismatchKind.FIELD_MISSING_IN_CODE,
                detail=f"declared in YAML ({yaml_schema.fields[field_name]}) but not in Pydantic model",
            ))
            continue

        if not in_yaml:
            mismatches.append(FieldMismatch(
                field_name=field_name,
                kind=MismatchKind.FIELD_MISSING_IN_YAML,
                detail=f"in Pydantic model ({pydantic_schema.fields[field_name]}) but not declared in YAML",
            ))
            continue

        # Both present — compare types
        yaml_type = yaml_schema.fields[field_name]
        code_type = pydantic_schema.fields[field_name]
        if yaml_type != "any" and code_type != "any" and yaml_type != code_type:
            mismatches.append(FieldMismatch(
                field_name=field_name,
                kind=MismatchKind.TYPE_MISMATCH,
                detail=f"YAML says '{yaml_type}', code says '{code_type}'",
            ))

        # Compare required status
        yaml_req = field_name in yaml_schema.required
        code_req = field_name in pydantic_schema.required
        if yaml_req and not code_req:
            mismatches.append(FieldMismatch(
                field_name=field_name,
                kind=MismatchKind.REQUIRED_MISMATCH,
                detail="required in YAML but optional in code",
            ))

    return mismatches


# ---------------------------------------------------------------------------
# Main check logic
# ---------------------------------------------------------------------------


TEMPLATES_DIR = Path(__file__).resolve().parent.parent / "app" / "config" / "templates"


def check_template(mapping: TemplateMapping) -> list[SchemaComparisonResult]:
    """Check one template file against its registered Pydantic models."""
    yaml_path = TEMPLATES_DIR / mapping.yaml_file
    with open(yaml_path) as f:
        template = yaml.safe_load(f)

    results: list[SchemaComparisonResult] = []

    # Check input_schema
    if mapping.input_model is not None and "input_schema" in template:
        yaml_norm = normalize_yaml_schema(template["input_schema"])
        code_norm = normalize_pydantic_schema(mapping.input_model)
        mismatches = compare_schemas(yaml_norm, code_norm)
        results.append(SchemaComparisonResult(
            template_file=mapping.yaml_file,
            schema_location="input_schema",
            model_name=mapping.input_model.__name__,
            mismatches=mismatches,
        ))

    # Check each step's result_schema
    steps_by_name = {s["name"]: s for s in template.get("steps", [])}
    for step_name, model_cls in mapping.step_models.items():
        step = steps_by_name.get(step_name)
        if step is None:
            results.append(SchemaComparisonResult(
                template_file=mapping.yaml_file,
                schema_location=f"step:{step_name}",
                model_name=model_cls.__name__,
                mismatches=[FieldMismatch(
                    field_name="<step>",
                    kind=MismatchKind.FIELD_MISSING_IN_YAML,
                    detail=f"step '{step_name}' not found in template YAML",
                )],
            ))
            continue

        result_schema = step.get("result_schema")
        if result_schema is None:
            results.append(SchemaComparisonResult(
                template_file=mapping.yaml_file,
                schema_location=f"step:{step_name}.result_schema",
                model_name=model_cls.__name__,
                mismatches=[FieldMismatch(
                    field_name="<schema>",
                    kind=MismatchKind.FIELD_MISSING_IN_YAML,
                    detail="no result_schema declared in template YAML",
                )],
            ))
            continue

        yaml_norm = normalize_yaml_schema(result_schema)
        code_norm = normalize_pydantic_schema(model_cls)
        mismatches = compare_schemas(yaml_norm, code_norm)
        results.append(SchemaComparisonResult(
            template_file=mapping.yaml_file,
            schema_location=f"step:{step_name}.result_schema",
            model_name=model_cls.__name__,
            mismatches=mismatches,
        ))

    return results


def check_all(
    template_filter: str | None = None,
) -> list[SchemaComparisonResult]:
    """Check all registered templates (or a single one by filename prefix)."""
    all_results: list[SchemaComparisonResult] = []
    for mapping in SCHEMA_REGISTRY:
        if template_filter and not mapping.yaml_file.startswith(template_filter):
            continue
        all_results.extend(check_template(mapping))
    return all_results


# ---------------------------------------------------------------------------
# Output formatting
# ---------------------------------------------------------------------------

_KIND_SYMBOLS = {
    MismatchKind.FIELD_MISSING_IN_CODE: "- CODE",
    MismatchKind.FIELD_MISSING_IN_YAML: "- YAML",
    MismatchKind.TYPE_MISMATCH: "~ TYPE",
    MismatchKind.REQUIRED_MISMATCH: "~ REQ ",
}


def print_report(results: list[SchemaComparisonResult], verbose: bool = False) -> int:
    """Print a human-readable report and return the count of schemas with mismatches."""
    total_schemas = len(results)
    failed_schemas = [r for r in results if not r.ok]

    if not failed_schemas:
        print(f"All {total_schemas} schemas consistent.")
        return 0

    # Group by template file
    by_file: dict[str, list[SchemaComparisonResult]] = {}
    for r in results:
        by_file.setdefault(r.template_file, []).append(r)

    for template_file, file_results in by_file.items():
        file_fails = [r for r in file_results if not r.ok]
        if not file_fails and not verbose:
            continue

        print(f"\n{template_file}")
        for r in file_results:
            status = "OK" if r.ok else "MISMATCH"
            count = f" ({len(r.mismatches)} issues)" if not r.ok else ""
            print(f"  {status} {r.schema_location} <-> {r.model_name}{count}")

            if not r.ok or verbose:
                for m in r.mismatches:
                    symbol = _KIND_SYMBOLS.get(m.kind, "?")
                    print(f"    {symbol}  {m.field_name}: {m.detail}")

    ok_count = total_schemas - len(failed_schemas)
    print(f"\n{ok_count}/{total_schemas} schemas consistent, "
          f"{len(failed_schemas)} with mismatches.")
    return len(failed_schemas)


# ---------------------------------------------------------------------------
# CLI entry point
# ---------------------------------------------------------------------------


def main() -> None:
    parser = argparse.ArgumentParser(description="Check template YAML schemas against Pydantic models")
    parser.add_argument("--verbose", "-v", action="store_true", help="Show all schemas, not just failures")
    parser.add_argument("--template", "-t", type=str, default=None, help="Filter to a single template by filename prefix")
    args = parser.parse_args()

    results = check_all(template_filter=args.template)
    fail_count = print_report(results, verbose=args.verbose)
    sys.exit(1 if fail_count > 0 else 0)


if __name__ == "__main__":
    main()
